internal/expr/parser.go

```diff
@@ -716,30 +742,150 @@ func (p *Parser) parseOutputExpr() (*outputExpr, bool, error) {
 	return nil, false, nil
 }
 
-// parseInputExpr parses an input expression of the form "$Type.name".
-func (p *Parser) parseInputExpr() (expression, bool, error) {
+// parseSliceInputExpr parses an input expression of the form "$Type[:]".
+func (p *Parser) parseSliceInputExpr() (*sliceInputExpr, bool, error) {
 	cp := p.save()
 	if !p.skipByte('$') {
 		return nil, false, nil
 	}
 
-	// Case 1: Slice range, "Type[:]".
 	if st, ok, err := p.parseSliceAccessor(); err != nil {
+		cp.restore()
```

Comment:
[Idea]: What do you think of restoring the `cp` using a defer clause checking the `err` value? It is more implicit but this way we don't have to remember to write it everywhere.

I am not saying to do it in this PR just asking for future changes.

Reply:
Yea, that would be better generally I think, there are a lot of places where it is repeated a ton.